drned/common/test_template.py

```diff
@@ -242,39 +241,59 @@ def _drned_single_set(device, init, fname, init_op, op, end_op, it, ordered):
         set_all_operation = True
         # Use upper scheme as default
         the_end = end_op
-        for tset_file in tset_files:
-            src,eop,sop = _drned_single_file(device, tset_file, op)
-            src_in_set = src_in_set or src
-            set_all_operation = set_all_operation and sop
-            # Get end operations from file, if present
-            if eop:
-                # File operations override
-                the_end = eop
-        if the_end == None:
-            # Default action is to first rollback to start, then to
-            # end, then rollback all commits, one by one
-            if set_all_operation:
-                the_end = ["rollback+0", "commit", "compare-config"]
-                extra_rollback = 1
+        try:
+            for tset_file in tset_files:
+                src,eop,sop = _drned_single_file(device, tset_file, op)
+                src_in_set = src_in_set or src
+                set_all_operation = set_all_operation and sop
+                # Get end operations from file, if present
+                if eop:
+                    # File operations override
+                    the_end = eop
+
+            if the_end == None:
+                # Default action is to first rollback to start, then to
+                # end, then rollback all commits, one by one
+                if set_all_operation:
+                    the_end = ["rollback+0", "commit", "compare-config"]
+                    extra_rollback = 1
+                else:
+                    the_end = []
+                    extra_rollback = 0
+                for i in reversed(range(commit_id_base, len(device.commit_id) + extra_rollback)):
+                    the_end += ["rollback+%d" % (i - commit_id_base), "commit", "compare-config"]
+            # Use the file function also for rollbacks
+            _drned_single_file(device, None, the_end,
+                               commit_id_base=commit_id_base)
+        except BaseException:
+            # Record the failed state and, if this is not the last
+            # state in the list, restore the device to its original
+            # configuration and continue testing. If this is the last
+            # test, re-raise the exception and trigger regular
+            # cleanup/error reporting.
+            device.failed_states.append(os.path.basename(tset_file))
+            if tset != tsets[-1]:
```

Comment:
Without treating the last failure special I got two rollbacks if the last state failed. As you say, another way of achieving the same thing is to drop the "standard" clean-up.

Reply:
I understand now what you mean by "two rollbacks". But is the `device.restore()` part in case of a failure needed at all? All `load` steps are done as `load override` for us (and if they were not, `device.restore()` does not help the situation, since it means that every `load` steps depends on the previous one, but in case of a failure this sequence of steps is lost anyway).

Reply:
Without a restore all following tests will fail. See the following snippet:
admin@ncs# devices device nc0 drned-xmnr transitions walk-states states [ drned-internal-asr9k-2-954 drned-internal-asr9k-3-875-pm-pbr-c-nexthop drned-internal-asr9k-2-664 drned-internal-asr9k-2-router-eigrp ] rollback true
Prepare the device
Test transition to drned-internal-asr9k-2-954
Test transition to drned-internal-asr9k-3-875-pm-pbr-c-nexthop
       failed
Test transition to drned-internal-asr9k-2-664
       failed
Test transition to drned-internal-asr9k-2-router-eigrp
       failed
Device cleanup
failure drned failed
admin@ncs#

The failures in drned-internal-asr9k-2-664 and drned-internal-asr9k-2-router-eigrp are reported in the compare_config() after the first rollback and is the same diff that triggered the error during drned-internal-asr9k-3-875-pm-pbr-c-nexthop.